modules/monitor/software/latency_monitor.cc

```diff
@@ -0,0 +1,106 @@
+/******************************************************************************
+ * Copyright 2019 The Apollo Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *****************************************************************************/
+
+#include "modules/monitor/software/latency_monitor.h"
+
+#include <algorithm>
+#include <memory>
+
+#include "cyber/common/log.h"
+#include "modules/common/adapters/adapter_gflags.h"
+#include "modules/monitor/common/monitor_manager.h"
+
+DEFINE_string(latency_monitor_name, "LatencyMonitor",
+              "Name of the latency monitor.");
+
+DEFINE_double(latency_monitor_interval, 2.0,
+              "Latency report interval in seconds.");
+
+DEFINE_double(latency_report_interval, 10.0,
+              "Latency report interval in seconds.");
+
+namespace apollo {
+namespace monitor {
+
+using apollo::common::LatencyRecordMap;
+using apollo::common::LatencyReport;
+using apollo::common::LatencyStat;
+using apollo::common::LatencyTrack;
+
+LatencyMonitor::LatencyMonitor()
+    : RecurrentRunner(FLAGS_latency_monitor_name,
+                      FLAGS_latency_monitor_interval) {}
+
+void LatencyMonitor::RunOnce(const double current_time) {
+  static auto reader =
+      MonitorManager::Instance()->CreateReader<LatencyRecordMap>(
+          FLAGS_latency_recording_topic);
+  reader->Observe();
+  const auto records = reader->GetLatestObserved();
+  if (records == nullptr || records->latency_records().empty()) {
+    return;
+  }
+
+  UpdateLatencyStat(records);
+
+  if (current_time - flush_time_ > FLAGS_latency_report_interval) {
+    flush_time_ = current_time;
+    PublishLatencyReport();
+  }
+}
+
+void LatencyMonitor::UpdateLatencyStat(
+    const std::shared_ptr<LatencyRecordMap>& records) {
+  if (latency_report_.stat_aggr().find(records->module_name()) ==
```

Comment:
Store `latency_report_.mutable_stat_aggr()` in a local pointer and reuse it. So the code is much simpler.